### DISCUSSION TITLE
Android packaging / build system

### DISCUSSION THREAD

**Discussion by DDRBoxman** [MusicPlayerDaemon/MPD#1917] [General]
I was looking at integrating media session into the Android app but I believe that it would first be worth the time to modernize how the app is build on the Android side. 

Android has the ability to package libraries as AAR files. These can include resources, native binaries, and java class files. I think it would be worthwhile to tweak the current build system to only build the native code and Bridge.java and package that as an AAR.
The rest of the project would be converted to a Gradle project so we can use more modern Android build tools, dependencies, and language features. This project would then pull in the AAR that was built so that the app can bundle the native MPD code. This change should make it easier to add functionality to the Android app going forward.

Let me know what you think!

Some quick info about AAR libraries:
https://developer.android.com/studio/projects/android-library

**Comment by MaxKellermann** [comment]
I don't see the point of Gradle. All you need to build an Android app is compile some Java (uh, it's a simple javac call), compile some native code (Meson does this well) and then put everything in a ZIP file (just a few shell calls).
Gradle adds hundreds of megabytes of Java bloat to do something that can be done with 1% of the CPU cycles.
You said Gradle enables the use of
- modern Android build tools
- dependencies
- language features

What Android build tools are impossible to be used without Gradle? What dependencies do you wish to have? What language features cannot be used without Gradle? What language, anyway? Java? Gradle isn't a Java compiler - it only invokes the Java compiler. Just like we do currently.

> Android has the ability to package libraries as AAR files. These can include resources, native binaries, and java class files.

I don't get that either. APK can also "include resources, native binaries, and java class files". In fact, that is exactly how MPD is packaged for Android currently.

**Reply by DDRBoxman** [reply]
The main dependency I'd like to look into using is the Jetpack Media3 library which makes managing metadata and playback control integration with the OS much easier and deals with backwards compatibility across older version of Android.

https://developer.android.com/guide/topics/media/session/mediasession

There is also a view binding feature that removes a lot of boiler plate around getting views from the xml into activities and fragments.

https://developer.android.com/topic/libraries/view-binding

The compose and material UI libraries are also pretty nice to have for building UIs.


The AAR is definitely not needed but seemed like a not terrible way to separate concerns. It's easy enough to just output the .so files to where gradle would expect them.

**Reply by MaxKellermann** [reply]
As much as I hate "modern" bloated tools like Gradle - if you're serious about improving MPD on Android, I would be very happy about that and I won't veto any build system change for the Android specific part, as long as it integrates well with MPD's current Meson build system (and of course as long as it complies with the license - i.e. don't link MPD with non-free stuff). There are other, leaner ways to do everything without Gradle, but I guess this would make you less productive and less happy with your work, so .... go for it.

**Reply by DDRBoxman** [reply]
Cool, I'll probably dig into this over the weekend.
